muted users: Add support to muting bots.

PIG208 opened the pull request:
We intentionally disallow muting bots previously upon a pending design decision in #16915.
This lifts that constraint.

Fixes #22693.

<!-- If the PR makes UI changes, always include one or more still screenshots to demonstrate your changes. If it seems helpful, add a screen capture of the new functionality as well.

Tooling tips: https://zulip.readthedocs.io/en/latest/tutorials/screenshot-and-gif-software.html
-->

**Screenshots and screen captures:**

<img width="653" alt="Screenshot 2023-06-13 122416" src="https://github.com/zulip/zulip/assets/39874143/caa5dc72-090d-426a-898b-6a5b387edb87">

<img width="656" alt="Screenshot 2023-06-13 122245" src="https://github.com/zulip/zulip/assets/39874143/cdf14445-4f70-4504-b1ff-6fe3b221ba83">


<details>
<summary>Self-review checklist</summary>

<!-- Prior to submitting a PR, follow our step-by-step guide to review your own code:
https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code -->

<!-- Once you create the PR, check off all the steps below that you have completed.
If any of these steps are not relevant or you have not completed, leave them unchecked.-->

- [ ] [Self-reviewed](https://zulip.readthedocs.io/en/latest/contributing/code-reviewing.html#how-to-review-code) the changes for clarity and maintainability
      (variable names, code reuse, readability, etc.).

Communicate decisions, questions, and potential concerns.

- [ ] Explains differences from previous plans (e.g., issue description).
- [ ] Highlights technical choices and bugs encountered.
- [ ] Calls out remaining decisions and concerns.
- [ ] Automated tests verify logic where appropriate.

Individual commits are ready for review (see [commit discipline](https://zulip.readthedocs.io/en/latest/contributing/commit-discipline.html)).

- [ ] Each commit is a coherent idea.
- [ ] Commit message(s) explain reasoning and motivation for changes.

Completed manual review and testing of the following:

- [ ] Visual appearance of the changes.
- [ ] Responsiveness and internationalization.
- [ ] Strings and tooltips.
- [ ] End-to-end functionality of buttons, interactions and flows.
- [ ] Corner cases, error conditions, and easily imagined bugs.
</details>

timabbott commented:
We need to do an API feature level bump and document the change on https://zulip.com/api/mute-user, since this contains an API change and in particular, mobile clients might want to show or not show the option to mute a bot user based on whether the server will accept that request.

While you're doing that tweak, can you also check for whether we mention the detail of whether bots can appear in the `muted_users` part of the events system docs.

PIG208 commented:
> This looks great in terms of the behavior!
> 
> We need to do an API feature level bump and document the change on https://zulip.com/api/mute-user, since this contains an API change and in particular, mobile clients might want to show or not show the option to mute a bot user based on whether the server will accept that request.
> 
> While you're doing that tweak, can you also check for whether we mention the detail of whether bots can appear in the `muted_users` part of the events system docs.

I didn't find this behavior mentioned anywhere in the documentation for the API or the event system part of it. Let's mention the old behavior in a `Changes` section and note that bot users are supported now.

PIG208 commented:
**Updates:**
- Bumped the feature level 188 skipping the two 7.x reserved levels.
- Added `Changes` sections for for `POST /users/me/muted_users/{muted_user_id}` and `DELETE /users/me/muted_users/{muted_user_id}` in the API documentation, and updated the changelog.

gnprice commented on zerver/openapi/zulip.yaml:
```diff
@@ -9023,6 +9026,9 @@ paths:
       description: |
         This endpoint unmutes a user.
 
+        **Changes**: Before Zulip 8.0 (feature level 188), this endpoint does not allow
+        unmuting bot users, indicated by returning an error with a "No such user" message.
```
The specific error message shouldn't be in the documentation, except as part of an example. That's because it will vary with translations, and so client code should never attempt to inspect it. (When there's information that client code is expected to want to extract from an error, we express it machine-readably in `code` and possible additional fields.)

Combining with the other comments above, this could become:

```suggestion
        **Changes**: Before Zulip 8.0 (feature level 188), it was an error to specify a bot user.
```

PIG208 replied:
That makes sense, thanks!

gnprice reviewed (commented):
Thanks for cleaning up this quirk! Comments below on the bit of API docs.

PIG208 commented:
The updated documentation:
<img width="548" alt="image" src="https://github.com/zulip/zulip/assets/39874143/387545b7-5970-4a8c-92d6-cd19b926d2ed">

timabbott commented:
This looks great, merged, thanks @PIG208!
